cmd/dockerd/daemon.go

```diff
@@ -484,7 +484,7 @@ func initRouter(s *apiserver.Server, d *daemon.Daemon, c *cluster.Cluster) {
 		image.NewRouter(d, decoder),
 		systemrouter.NewRouter(d, c),
 		volume.NewRouter(d),
-		build.NewRouter(dockerfile.NewBuildManager(d)),
+		build.NewRouter(manager.NewBuildManager(d, d), d),
```

Comment:
Thats not quite what I had in mind. I meant adding `tagSquashBackend` as the second parameter to `NewRouter`, not `NewBuildManager`.

Looking over the code I'm not sure how well it applies to squash at all as it would be much cleaner for the builder to just return image ID.  Squashing does look like a build step, although a hacky one and not similar to the rest of them.

nit: Otoh, `HasExperimental` is a daemon concept that builder(and probably even router) should not know about. I wouldn't think it would be too bad for the `SquashImage()` just to error, or to fail early it could call `SqushImageSupported() bool`.

Also, we should open up an issue to discuss if we want to continue with squash or deprecate it and never move it out of experimental. Multi stage covers the same use cases, and `RUN --mount` would let squash differences between arbitrary directory structures.